src/main/java/com/endava/internship/service/UserServiceImpl.java

```diff
@@ -2,68 +2,105 @@
 
     @Override
-    public List<String> getFirstNamesReverseSorted(List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+    public List<String> getFirstNamesReverseSorted(final List<User> users) {
+        return users.stream()
+                .map(User::getFirstName)
+                .sorted(Comparator.reverseOrder())
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<User> sortByAgeDescAndNameAsc(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .sorted(Comparator.comparingInt(User::getAge).reversed()
+                        .thenComparing(User::getFirstName))
+                .collect(Collectors.toList());
     }
 
     @Override
     public List<Privilege> getAllDistinctPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .flatMap(user -> user.getPrivileges().stream())
+                .distinct()
+                .collect(Collectors.toList());
     }
 
     @Override
     public Optional<User> getUpdateUserWithAgeHigherThan(final List<User> users, final int age) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .filter(user -> user.getPrivileges().contains(Privilege.UPDATE) && user.getAge() > age)
+                .findAny();
     }
 
     @Override
     public Map<Integer, List<User>> groupByCountOfPrivileges(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .collect(Collectors.groupingBy(user -> user.getPrivileges().size()));
     }
 
     @Override
     public double getAverageAgeForUsers(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        return users.stream()
+                .mapToDouble(User::getAge)
+                .average().orElse(-1);
     }
 
     @Override
     public Optional<String> getMostFrequentLastName(final List<User> users) {
-        throw new UnsupportedOperationException("Not implemented");
+        final var map = getNumberOfLastNames(users);
+        final var list = map.entrySet().stream()
+                .filter(stringLongEntry -> stringLongEntry.getValue().equals(map.values().stream().max(Long::compare).get()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.toList());
+        return list.size() > 1 ? Optional.empty() : list.stream().findFirst();
     }
 
     @Override
     public List<User> filterBy(final List<User> users, final Predicate<User>... predicates) {
-        throw new UnsupportedOperationException("Not implemented");
+        if (predicates.length < 1) {
+            return users;
+        }
+        var stream = users.stream();
+        for (Predicate<User> predicate : predicates) {
```

Comment:
you can define a predicate
you will have Predicate<User> allThePredicates = ...
and then filter by them

Reply:
Done